sycl/include/sycl/marray.hpp

```diff
@@ -85,20 +72,44 @@ template <typename Type, std::size_t NumElements> class marray {
   static constexpr auto FlattenMArrayArgHelper(const T &A) {
     return std::array<DataT, 1>{static_cast<DataT>(A)};
   }
-  template <typename DataT, typename T> struct FlattenMArrayArg {
-    constexpr auto operator()(const T &A) const {
-      return FlattenMArrayArgHelper<DataT>(A);
-    }
-  };
+};
 
-  // Alias for shortening the marray arguments to array converter.
-  template <typename DataT, typename... ArgTN>
-  using MArrayArgArrayCreator =
-      detail::ArrayCreator<DataT, FlattenMArrayArg, ArgTN...>;
+template <typename DataT, typename T> struct FlattenMArrayArg {
+  constexpr auto operator()(const T &A) const {
+    return HelperFlattenMArrayArg::FlattenMArrayArgHelper<DataT>(A);
+  }
+};
+
+// Alias for shortening the marray arguments to array converter.
+template <typename DataT, typename... ArgTN>
+using MArrayArgArrayCreator =
+    detail::ArrayCreator<DataT, FlattenMArrayArg, ArgTN...>;
+
+} // namespace detail
+
+/// Provides a cross-platform math array class template that works on
+/// SYCL devices as well as in host C++ code.
+///
+/// \ingroup sycl_api
+template <typename Type, std::size_t NumElements> class marray {
+  using DataT = Type;
 
-  // FIXME: Other marray specializations needs to be a friend to access MData.
-  //        If the subscript operator is made constexpr this can be removed.
+public:
+  using value_type = Type;
+  using reference = Type &;
+  using const_reference = const Type &;
+  using iterator = Type *;
+  using const_iterator = const Type *;
+
+private:
+  value_type MData[NumElements];
+
+  /// FIXME: If the subscript operator is made constexpr these can be removed.
+  // Other marray specializations needs to be a friend to access MData.
```

Comment:
Nit: I don't mind changing it so that `FIXME` is only the actionable part of the comment, but I would then prefer it to be the later part of the comment:
```suggestion
  // Other marray specializations needs to be a friend to access MData.
  // FIXME: If the subscript operator is made constexpr these can be removed.
```